Method overload object type

vmuriart opened the pull request:
### What does this implement/fix? Explain your changes.

Fixes Method object overloading without breaking changes introduced by #131 and #151.

I added the tests/examples from #131 to the [tests](https://github.com/pythonnet/pythonnet/blob/6afc9e684856fca988c6d5260ee2d6de3735b7cf/src/tests/test_method.py#L748-L771) to ensure that the behavior didn't get reverted.

### Does this close any currently open issues?

Closes #203 

### Any other comments?

This patch fixes the issue from #203, but `MethodBinder.Bind` needs rewriting since its basically been patched over and over again to fix edge cases. I didn't want to do it as part of this `pull_request` since it would affect more than just fixing #203.

### Checklist

Check all those that are applicable and complete.

-   [x] Make sure to include one or more tests for your change
-   [x] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)

denfromufa commented:
@vmuriart what happens with multiple arguments, when overloaded methods have `object` argument?

vmuriart commented:
@denfromufa good question. No idea.

`foo(int a, object b)`
`foo(object a, int b)`
`foo(object a, object b)`
`foo(int a, int b)`

These 4 functions look enough to cover the possibilities or do I need more?

denfromufa commented:
we need one super-loaded method which covers all these cases from this meta-issue :)

https://github.com/pythonnet/pythonnet/issues/265

vmuriart commented:
That sounds much broader than the scope of this `pull_request`. 
I'm planning on refactoring `methodbinder` to clean up alot of what is going in there.

denfromufa commented:
@vmuriart I understand, your example looks good. My suggestion was not serious anyway. But we may hit issues when dealing with sub-classes and ambiguous cases in the overload. IMO, pythonnet should be very strict by default in method resolution. See examples how C# compiler had to break method overloading in .NET 4.0:

http://csharpindepth.com/Articles/General/Overloading.aspx

vmuriart commented:
@denfromufa No worries. My goal is to restore similar behavior to what `pythonnet==2.0.0` had irt method objects. I had to freeze the version of pythonnet I was using to `v2.0` because of it.
I'm ok w switching to a stricter method resolution, as long as it is part of `v3.0` 😄

vmuriart commented:
@denfromufa Multiple arguments with objects works as expected.

denfromufa commented:
@vmuriart no failure cases in overloads to make sure that they fail. No strings in overloads, although you use strings as arguments.

vmuriart commented:
I think I understood your request correctly. I added a test to check for failure/exception.

denfromufa commented:
Can you add more overloaded methods such as:

```
public static string TestOverloadedObjectTwo(string a, string b){return
"Got string-string";}
```

Which one would the method resolution pick object or string? Does the order
of defined methods matter?

Wrong result would be much worse than flagging ambiguity or failure in
method resolution.

On Wed, Feb 15, 2017, 1:26 AM Victor Uriarte <notifications@github.com>
wrote:

> I think I understood your request correctly. I added a test to check for
> failure/exception.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pythonnet/pythonnet/pull/377#issuecomment-279936453>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHgZ5ZWYSupa6rOYtu-trQflb6-85xhrks5rcqiogaJpZM4MBRiU>
> .
>

vmuriart commented:
@denfromufa I think you are asking for [these tests](https://github.com/pythonnet/pythonnet/blob/6afc9e684856fca988c6d5260ee2d6de3735b7cf/src/tests/test_method.py#L748-L771) I added them on a [separate commit](https://github.com/pythonnet/pythonnet/commit/ce40fb1a72e1e10ec8428bdd55bab7c08e2866ca). 

Earlier when I was working on this pull request I used those tests to make sure I wasn't reverting the behavior you are describing. On earlier implementations, I reverted the behavior so those tests kept me in check.

The string-string check is analogous to the [int-int](https://github.com/pythonnet/pythonnet/pull/377/files#diff-1eb23401deaa91056741630bfbed1db7R134) check vs [obj-obj](https://github.com/pythonnet/pythonnet/pull/377/files#diff-1eb23401deaa91056741630bfbed1db7R149) check

vmuriart commented:
@denfromufa `int` are also a subclass of `System.Object`.

```
System.Object
   System.ValueType
      System.Int32
```

denfromufa commented:
yes, int is also object subclass. However you use int in C# overloaded methods, while not using strings. Let me give some examples.

denfromufa commented:
@vmuriart here is what I suggested to add. I cannot test locally, due to #385 

```CSHARP
public static string TestOverloadedObject(object o)
        {
            return "Got object";
        }

        public static string TestOverloadedObjectTwo(int a, int b)
        {
            return "Got int-int";
        }

        public static string TestOverloadedObjectTwo(string a, string b)
        {
            return "Got string-string";
        }

        public static string TestOverloadedObjectTwo(string a, int b)
        {
            return "Got string-int";
        }

        public static string TestOverloadedObjectTwo(string a, object b)
        {
            return "Got string-object";
        }

        public static string TestOverloadedObjectTwo(int a, object b)
        {
            return "Got int-object";
        }

        public static string TestOverloadedObjectTwo(object a, int b)
        {
            return "Got object-int";
        }

        public static string TestOverloadedObjectTwo(object a, object b)
        {
            return "Got object-object";
        }

        public static string TestOverloadedObjectThree(object a, int b)
        {
            return "Got object-int";
        }

        public static string TestOverloadedObjectThree(int a, object b)
        {
            return "Got int-object";
        }
```

```
def test_object_in_multiparam():
    """Test method with object multiparams behaves"""

    res = MethodTest.TestOverloadedObjectTwo(5, 5)
    assert res == "Got int-int"

    res = MethodTest.TestOverloadedObjectTwo(5, "foo")
    assert res == "Got int-object"

    res = MethodTest.TestOverloadedObjectTwo("foo", 5)
    assert res == "Got object-int"

    res = MethodTest.TestOverloadedObjectTwo("foo", "bar")
    assert res == "Got object-object"

    res = MethodTest.TestOverloadedObjectTwo("foo", System.Object)
    assert res == "Got string-object"
   
    res = MethodTest.TestOverloadedObjectTwo("foo", "bar")
    assert res == "Got string-string"

    res = MethodTest.TestOverloadedObjectTwo("foo", 5)
    assert res == "Got string-int"


```

vmuriart commented:
ok, I'll add these tomorrow

vmuriart commented:
@denfromufa Added your tests, but had to modify remove a couple of the conditions I had added earlier since the function signatures had changed.

denfromufa commented on src/runtime/methodbinder.cs:
```diff
@@ -375,7 +375,7 @@ internal Binding Bind(IntPtr inst, IntPtr args, IntPtr kw, MethodBase info, Meth
                             {
                                 var typematch = false;
-                                if (pi[n].ParameterType != clrtype)
+                                if (pi[n].ParameterType != typeof(object) && pi[n].ParameterType != clrtype)
```
can you add brackets around each logical?

denfromufa commented:
@vmuriart I left 2 comments inline with the code

vmuriart commented:
I'll review on Tuesday. Though the parenthesis seem superfluous since it
isn't consistent throughout.

On Fri, Feb 17, 2017 at 18:56 denfromufa <notifications@github.com> wrote:

> @vmuriart <https://github.com/vmuriart> I left 2 comments inline with the
> code
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pythonnet/pythonnet/pull/377#issuecomment-280816555>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMr87EoUuLcuTE5k0mXi3kl9MNwyHCvPks5rdl3GgaJpZM4MBRiU>
> .
>

denfromufa commented:
yes, brackets is very minor. But I feel the same way as this person:

http://stackoverflow.com/a/23816736/2230844

On Fri, Feb 17, 2017 at 9:28 PM, Victor Uriarte <notifications@github.com>
wrote:

> I'll review on Tuesday. Though the parenthesis seem superfluous since it
> isn't consistent throughout.
>
> On Fri, Feb 17, 2017 at 18:56 denfromufa <notifications@github.com> wrote:
>
> > @vmuriart <https://github.com/vmuriart> I left 2 comments inline with
> the
> > code
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/pythonnet/pythonnet/pull/377#issuecomment-280816555
> >,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AMr87EoUuLcuTE5k0mXi3kl9MNwyHCvPks5rdl3GgaJpZM4MBRiU>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pythonnet/pythonnet/pull/377#issuecomment-280818253>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHgZ5QThkeZd3Re4qve47oNeO1lnknICks5rdmVVgaJpZM4MBRiU>
> .
>

vmuriart commented:
@denfromufa Added the parenthesis. Good to squash-merge?

denfromufa commented:
you did not address the second comment. @vmuriart will this break if you add:
```
public static string TestOverloadedObjectTwo(int a, string b)
        {
            return "Got int-string";
        }
```

vmuriart commented:
There was only a single comment. 

![image](https://cloud.githubusercontent.com/assets/13303020/23145145/b0b399de-f789-11e6-92a6-00cb40b072c1.png)

And the answer is no. It simply changes the result of which method is used as it will pick the closest match it can find before it picked `object`.

denfromufa commented:
so can you add it? because testing closest match is important.

so you don't see this?

![image](https://cloud.githubusercontent.com/assets/7870949/23145335/2f8b5bc6-f793-11e6-811a-3432286e374e.png)

vmuriart commented:
It's not there. It even says on your screen shot that its `pending`.

vmuriart commented on src/tests/test_method.py:
```diff
@@ -769,3 +769,50 @@ def test_wrong_overload():
+    res = MethodTest.TestOverloadedObjectTwo(5, 5)
+    assert res == "Got int-int"
+
+    res = MethodTest.TestOverloadedObjectTwo(5, "foo")
```
Now its there.

vmuriart commented:
Added test and its passing as well.

denfromufa commented on src/testing/methodtest.cs:
```diff
@@ -116,6 +116,71 @@ public static int[] TestOverloadedParams(int v, int[] args)
+            return "Got object-int";
+        }
+
+        public static string TestOverloadedObjectTwo(object a, object b)
```
you don't test this method below

vmuriart replied:
because you changed the test for it above. I re-added a test for it.

vmuriart commented:
@denfromufa test added and passing as well.

denfromufa commented:
@vmuriart I'm still surprised that this does not break ;) i approve on my side. in the future we should definitely improve testing all these combinations of arguments by using hypothesis or parametrizing pytest. Let's release v2.3.0.dev1 version after this merge.

@filmor @tonyroberts please review!

vmuriart commented:
@denfromufa I didn't want to introduce `parametrizing` tests yet. I figured I'd slowly introduce `py.test` features instead of all at once.
